portal-kernel/src/com/liferay/portal/kernel/portlet/render/PortletRenderUtil.java

```diff
@@ -0,0 +1,592 @@
+	}
+
+	private static Collection<String> _getStaticURLs(
+		HttpServletRequest httpServletRequest,
+		Collection<PortletResourceAccessor> portletResourceAccessors,
+		Collection<Portlet> portlets, Set<String> visitedURLs) {
+
+		List<String> urls = new ArrayList<>();
+
+		ThemeDisplay themeDisplay =
+			(ThemeDisplay)httpServletRequest.getAttribute(
+				WebKeys.THEME_DISPLAY);
+
+		for (Portlet portlet : portlets) {
+			Portlet rootPortlet = portlet.getRootPortlet();
+
+			for (PortletResourceAccessor portletResourceAccessor :
+					portletResourceAccessors) {
+
+				String contextPath = null;
+
+				if (portletResourceAccessor.isPortalResource()) {
+					contextPath = PortalUtil.getPathContext();
+				}
+				else {
+					contextPath =
+						PortalUtil.getPathProxy() + portlet.getContextPath();
+				}
+
+				Collection<String> portletResources =
+					portletResourceAccessor.get(portlet);
+
+				for (String portletResource : portletResources) {
+					boolean module = false;
+
+					if (portletResource.startsWith("module:")) {
+						module = true;
+
+						portletResource = portletResource.substring(7);
+					}
+
+					if (!HttpComponentsUtil.hasProtocol(portletResource)) {
+						portletResource = PortalUtil.getStaticResourceURL(
+							httpServletRequest, contextPath + portletResource,
+							rootPortlet.getTimestamp());
+					}
+
+					if (!portletResource.contains(Http.PROTOCOL_DELIMITER)) {
+						String cdnBaseURL = themeDisplay.getCDNBaseURL();
+
+						portletResource = cdnBaseURL.concat(portletResource);
+					}
+
+					if (module) {
+						portletResource = "module:" + portletResource;
+					}
+
+					if (visitedURLs.contains(portletResource)) {
+						continue;
+					}
+
+					visitedURLs.add(portletResource);
+
+					urls.add(portletResource);
+				}
+			}
+		}
+
+		return urls;
+	}
+
+	private static Collection<String> _getURLs(
+		HttpServletRequest httpServletRequest,
+		Collection<PortletResourceAccessor> portletResourceAccessors,
+		Collection<Portlet> portlets, URLType urlType) {
+
+		boolean fastLoad;
+		Predicate<String> predicate = null;
+
+		ThemeDisplay themeDisplay =
+			(ThemeDisplay)httpServletRequest.getAttribute(
+				WebKeys.THEME_DISPLAY);
+
+		if (urlType == URLType.CSS) {
+			fastLoad = themeDisplay.isThemeCssFastLoad();
+		}
+		else if (urlType == URLType.JAVASCRIPT) {
+			fastLoad = themeDisplay.isThemeJsFastLoad();
+
+			predicate = resource -> !themeDisplay.isIncludedJs(resource);
+		}
+		else {
+			throw new UnsupportedOperationException(
+				"Unsupported URL type " + urlType);
+		}
+
+		Set<String> visitedURLs = (Set<String>)httpServletRequest.getAttribute(
+			WebKeys.PORTLET_RESOURCE_STATIC_URLS);
+
+		if (visitedURLs == null) {
+			visitedURLs = new LinkedHashSet<>();
+
+			httpServletRequest.setAttribute(
+				WebKeys.PORTLET_RESOURCE_STATIC_URLS, visitedURLs);
+		}
+
+		if (fastLoad) {
+			Theme theme = themeDisplay.getTheme();
+
+			return _getComboServletURLs(
```

Comment:
Are early returns okay to use in portal? For some reason I remember bchan or someone not liking them... I could be widely off here though

Reply:
It's now the blessed pattern. "We" even prefer using "if + return then more code" than "if + return else more code"